File: Set 8/Number of Islands/IslandsCount.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(vector<vector<int>>& grid, int i, int j, int n, int m) {
    if (i < 0 || j < 0 || i >= n || j >= m || grid[i][j] == 0) {
        return;
    }
    grid[i][j] = 0; 
    dfs(grid, i + 1, j, n, m); 
    dfs(grid, i - 1, j, n, m); 
    dfs(grid, i, j + 1, n, m); 
    dfs(grid, i, j - 1, n, m); 
    dfs(grid, i + 1, j + 1, n, m); 
    dfs(grid, i - 1, j - 1, n, m); 
    dfs(grid, i + 1, j - 1, n, m); 
    dfs(grid, i - 1, j + 1, n, m); 
}

int countIslands(vector<vector<int>>& grid) {
    int n = grid.size();
    int m = grid[0].size();
    int count = 0;

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (grid[i][j] == 1) {
                count++;
                dfs(grid, i, j, n, m); 
            }
        }
    }

    return count;
}

int main() {
    int t;
    cout << "Enter number of test cases: ";
    cin >> t;

    while (t--) {
        int n, m;
        cout << "Enter dimensions of the grid (rows and columns): ";
        cin >> n >> m;
        vector<vector<int>> grid(n, vector<int>(m));
        cout << "Enter the grid (0s and 1s):\n";
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                cin >> grid[i][j];
            }
        }

        cout << "Number of islands: " << countIslands(grid) << endl;
    }

    return 0;
}

File: Set 8/Number of Islands/IslandsCount.py
def dfs(grid, i, j, n, m):
    if i < 0 or j < 0 or i >= n or j >= m or grid[i][j] == 0:
        return
    grid[i][j] = 0  
    dfs(grid, i + 1, j, n, m)  # D
    dfs(grid, i - 1, j, n, m)  # U
    dfs(grid, i, j + 1, n, m)  # R
    dfs(grid, i, j - 1, n, m)  # L
    dfs(grid, i + 1, j + 1, n, m)  # BR
    dfs(grid, i - 1, j - 1, n, m)  # TL
    dfs(grid, i + 1, j - 1, n, m)  # BL
    dfs(grid, i - 1, j + 1, n, m)  # TR

def count_islands(grid):
    n, m = len(grid), len(grid[0])
    count = 0

    for i in range(n):
        for j in range(m):
            if grid[i][j] == 1:
                count += 1
                dfs(grid, i, j, n, m)  
    return count

def main():
    t = int(input("Enter number of test cases: "))
    for _ in range(t):
        n, m = map(int, input("Enter dimensions of the grid (rows and columns): ").split())
        print("Enter the grid (0s and 1s):")
        grid = [list(map(int, input().split())) for _ in range(n)]

        print("Number of islands:", count_islands(grid))

if __name__ == "__main__":
    main()
